integration/gethnetwork/geth_network.go

```diff
@@ -0,0 +1,140 @@
+package gethnetwork
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+const (
+	genesisFileName = "genesis.json"
+	genesisConfig   = "{\n  \"config\": {\n    \"chainId\": 777,\n    \"homesteadBlock\": 0,\n    \"eip150Block\": 0,\n    \"eip155Block\": 0,\n    \"eip158Block\": 0,\n    \"byzantiumBlock\": 0,\n    \"constantinopleBlock\": 0,\n    \"petersburgBlock\": 0,\n    \"istanbulBlock\": 0,\n    \"berlinBlock\": 0,\n    \"londonBlock\": 0\n  },\n  \"alloc\": {\n    \"0x0000000000000000000000000000000000000001\": {\n      \"balance\": \"111111111\"\n    }\n  },\n  \"coinbase\": \"0x0000000000000000000000000000000000000000\",\n  \"difficulty\": \"0x20000\",\n  \"extraData\": \"\",\n  \"gasLimit\": \"0x2fefd8\",\n  \"nonce\": \"0x0000000000000042\",\n  \"mixhash\": \"0x0000000000000000000000000000000000000000000000000000000000000000\",\n  \"parentHash\": \"0x0000000000000000000000000000000000000000000000000000000000000000\",\n  \"timestamp\": \"0x00\"\n}"
+	nodeFolderName  = string(os.PathSeparator) + "node%d_datadir" + string(os.PathSeparator)
+	ipcFileName     = "geth.ipc"
+	tempDirPrefix   = "geth_nodes"
+
+	dataDirFlag = "--datadir"
+	execFlag    = "--exec"
+	portFlag    = "--port"
+
+	addPeerCmd = "admin.addPeer(%s)"
+	attachCmd  = "attach"
+	enodeCmd   = "admin.nodeInfo.enode"
+	initCmd    = "init"
+)
+
+// GethNetwork is a network of Geth nodes, built using the provided Geth binary.
+type GethNetwork struct {
+	gethBinaryPath  string
+	genesisFilePath string
+	dataDirs        []string
+}
+
+// NewGethNetwork using the provided Geth binary to create a private Ethereum network with numNodes Geth nodes.
+func NewGethNetwork(gethBinaryPath string, numNodes int) GethNetwork {
+	nodesDir, err := ioutil.TempDir("", tempDirPrefix)
+	if err != nil {
+		panic(err)
+	}
+
+	// We write out the `genesis.json` file to be used by the network.
+	genesisFilePath := nodesDir + string(os.PathSeparator) + genesisFileName
+	err = os.WriteFile(genesisFilePath, []byte(genesisConfig), 0o600)
+	if err != nil {
+		panic(err)
+	}
+
+	// Each Geth node needs its own data directory.
+	dataDirs := make([]string, numNodes)
+	for i := 0; i < numNodes; i++ {
+		dataDirs[i] = fmt.Sprintf(nodesDir+nodeFolderName, i+1)
+	}
+
+	network := GethNetwork{
+		gethBinaryPath:  gethBinaryPath,
+		genesisFilePath: genesisFilePath,
+		dataDirs:        dataDirs,
+	}
+
+	for i, dataDir := range dataDirs {
+		_ = os.Remove(dataDir + ipcFileName) // We delete leftover IPC files from previous runs.
+		go network.createGethNode(dataDir, 30303+i)
```

Comment:
Are we piping the logs somewhere ? I may have missed that.

Should we surface errors here ? Like if the node dies for some reason.